### PR TITLE
fix(console): should not display unsaved alert on item deleted

### PR DESCRIPTION
**Pull request by xiaoyijun**
## Summary

- fix: replace `useEffect` with `useLayoutEffect` in `UnsavedChangesAlert` for we need to mutate the logic synchronously before we navigate to another page.
- refactor: extract `DeleteForm` into a commonly used `ResourceDeleteConfirmModal` component.
- refactor: replace `DeleteForm` on user-details page with `ConfirmModal`.
- fix: LOG-3609 - do not call `setState` when navigate to another page on `Application` or `ApiResource` deleted (replace `try...finally` block with `try...catch` block).
- feat: support `confirmDisable` and `loading` state in `ConfirmModal`.

## Testing

Tested in the Admin Console.

### CODE REVIEW

File: packages/console/src/pages/ApiResourceDetails/components/DeleteForm/index.tsx [hunk 1]
```diff
@@ -13,7 +13,7 @@ import * as styles from './index.module.scss';
 type Props = {
   id: string;
   name: string;
-  onClose: () => void;
+  onClose: (hasDeleted: boolean) => void;
```
Comment (simeng-li) [inline packages/console/src/pages/ApiResourceDetails/components/DeleteForm/index.tsx:16]: make it optional, default false?

**Review by wangsijie — APPROVED** [review]

File: packages/console/src/pages/ApplicationDetails/index.tsx [hunk 2]
```diff
@@ -97,6 +98,23 @@ const ApplicationDetails = () => {
...
+    try {
+      await api.delete(`/api/applications/${data.id}`);
+      setIsDeleted(true);
+      setIsDeleting(false);
+      setIsDeleteFormOpen(false);
+      toast.success(t('application_details.application_deleted', { name: data.name }));
+      navigate(`/applications`);
+    } catch {
+      setIsDeleting(false);
+    }
```
Comment (simeng-li) [inline packages/console/src/pages/ApplicationDetails/index.tsx:115]: nit: just an insight. feels like a good place to extract it as a hook or maybe that takes the delete endpoint and a callback URL as props.

**Review by simeng-li — COMMENTED** [review]
thanks for the efforts. Never get tired of seeing tons of code being removed.

File: packages/console/src/components/ConfirmModal/index.tsx [hunk 3]
```diff
@@ -43,7 +47,13 @@ const ConfirmModal = ({
         footer={
           <>
             <Button type="outline" title={cancelButtonText} onClick={onCancel} />
-            <Button type={confirmButtonType} title={confirmButtonText} onClick={onConfirm} />
+            <Button
+              type={confirmButtonType}
+              title={confirmButtonText}
+              disabled={isConfirmDisabled}
+              isLoading={isLoading}
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply packages/console/src/components/ConfirmModal/index.tsx:54]: Just keep the original deletion behavior before refactoring.
The deletion process may fail, and this prop is optional, not a force requirement.

File: packages/console/src/components/ResourceDeleteConfirmModal/index.tsx [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      className={styles.content}
+      onCancel={onCancel}
+      onConfirm={onConfirm}
+    >
+      <div className={styles.description}>
+        <Trans components={{ span: <span className={styles.highlight} /> }}>
+          {t(messageTemplate, { name })}
+        </Trans>
+      </div>
+      <TextInput
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply packages/console/src/components/ResourceDeleteConfirmModal/index.tsx:49]: This logic is used in `Application` and `ApiResource`; other situations can use `ConfirmModal` instead. (e.g., User deletion on user details page).

File: packages/console/src/pages/ApplicationDetails/index.tsx (hunk 2 above)
Comment (xiaoyijun) [reply packages/console/src/pages/ApplicationDetails/index.tsx:115]: Maybe good to extract the CRUD operations & states into a hook in the future.
Leave it to LOG-3638.

File: packages/console/src/pages/ApiResourceDetails/index.tsx [hunk 5]
```diff
@@ -18,6 +17,7 @@ import CopyToClipboard from '@/components/CopyToClipboard';
 import DetailsSkeleton from '@/components/DetailsSkeleton';
 import FormField from '@/components/FormField';
 import LinkButton from '@/components/LinkButton';
+import ResourceDeleteConfirmModal from '@/components/ResourceDeleteConfirmModal';
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply packages/console/src/pages/ApiResourceDetails/index.tsx:20]: Hard to find a more suitable name.
This component is only used in `Application` and `ApiResource`; they're all 'resources'.

File: packages/console/src/components/ConfirmModal/index.tsx [hunk 6]
```diff
@@ -17,6 +17,8 @@ export type ConfirmModalProps = {
   confirmButtonText?: AdminConsoleKey;
   cancelButtonText?: AdminConsoleKey;
   isOpen: boolean;
+  isConfirmDisabled?: boolean;
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply packages/console/src/components/ConfirmModal/index.tsx]: Updated

**Comment by xiaoyijun** [comment]
Close for this PR contains too many things.

File: packages/console/src/components/DeleteConfirmModal/index.tsx [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+import ConfirmModal from '@/components/ConfirmModal';
+import TextInput from '@/components/TextInput';
+
+type Props = {
+  isOpen: boolean;
+  isLoading?: boolean;
+  children: ReactNode;
+  expectedInput?: string;
+  inputPlaceholder?: string;
```
Comment (simeng-li) [inline packages/console/src/components/DeleteConfirmModal/index.tsx:11]: nit: not sure, but do you think wrapping them up into a typed object could be more self-contained? 
e.g.
```
type Captcha = {
 captchaValue: string;
 placeholder: string;
}


{ captcha && <TextInput />}
```

**Review by simeng-li — COMMENTED** [review]
overall LGTM. please make sure to test well before merging, as we do not have ut in AC.

File: packages/console/src/components/DeleteConfirmModal/index.tsx (hunk 7 above)
Comment (demonzoo) [reply packages/console/src/components/DeleteConfirmModal/index.tsx:11]: I'm good with both

Comment (xiaoyijun) [reply packages/console/src/components/DeleteConfirmModal/index.tsx:11]: I thought the placeholder could be optional, so it's ok to keep what it is.

File: packages/console/src/components/DeleteConfirmModal/index.tsx [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  onConfirm,
+}: Props) => {
+  const [input, setInput] = useState('');
+  const inputMismatched = input !== expectedInput;
+
+  return (
+    <ConfirmModal
+      isOpen={isOpen}
+      isLoading={isLoading}
+      isConfirmButtonDisabled={Boolean(expectedInput) && inputMismatched}
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply packages/console/src/components/DeleteConfirmModal/index.tsx]: updated.

**Pull request merged**